docs/monitoring.md

```diff
@@ -1,32 +1,83 @@
 # Monitoring
 
-## Monitoring container resource usage
+## Prometheus
 
-When run in docker, a few additional containers are started up:
+To monitor RPC return values over time, a Prometheus data exporter can be connected
+to any Bitcoin Tank and configured to scrape any available RPC results.
 
-* CAdvisor (container Monitoring)
-* Prometheus (log scraper)
-* Grafana (graphing/dashboard tool)
+The `bitcoin-exporter` image is defined in `resources/images/exporter` and
+maintained in the BitcoinDevProject dockerhub organization. To add the exporter
+in the Tank pod with Bitcoin Core add the `"exporter"` key to the node in the graphml file:
 
-## CAdvisor
+```xml
+    <node id="0">
+        <data key="version">27.0</data>
+        <data key="exporter">true</data>
+    </node>
+```
 
-CAdvisor needs no additional setup, and can be accessed from the docker host at
-localhost:8080
+The default metrics are defined in the `bitcoin-exporter` image:
+- Block count
+- Number of inbound peers
+- Number of outbound peers
+- Mempool size (# of TXs)
 
-## Prometheus
+Metrics can be configured by setting a `"metrics"` key to the node in the graphml file.
+The metrics value is a space-separated list of labels, RPC commands with arguments, and
+JSON keys to resolve the desired data:
+
+```
+label=method(arguments)[JSON result key][...]
+```
+
+For example, the default metrics listed above are defined as:
+
+```xml
+    <node id="0">
+        <data key="version">27.0</data>
+        <data key="exporter">true</data>
+        <data key="metrics">blocks=getblockcount() inbounds=getnetworkinfo()["connections_in"] outbounds=getnetworkinfo()["connections_in"] mempool_size=getmempoolinfo()["size"]</data>
+    </node>
+```
+
+The data can be retrieved from the Prometheus exporter on port `9332`, example:
 
-Prometheus should also not need any additional setup, and can be accessed from
-the docker host at localhost:9090
+```
+# HELP blocks getblockcount()
+# TYPE blocks gauge
+blocks 704.0
+# HELP inbounds getnetworkinfo()["connections_in"]
+# TYPE inbounds gauge
+inbounds 0.0
+# HELP outbounds getnetworkinfo()["connections_in"]
+# TYPE outbounds gauge
+outbounds 0.0
+# HELP mempool_size getmempoolinfo()["size"]
+# TYPE mempool_size gauge
+mempool_size 0.0
+```
 
 ## Grafana
 
-Grafana is provisioned with a single default dashboard, but alternative
-dashboards can be added or created.
+Data from Prometheus exporters can be collected and fed into Grafana for a
+web-based interface.
+
+### Install logging infrastructure
+
+First make sure you have `helm` installed, then run the `install_logging` script:
+
+```bash
+resources/scripts/install_logging.sh
+```
+
+To forward port `3000` and view the Grafana dashboard run the `connect_logging` script:
+
+```bash
+resources/scripts/connect_logging.sh
+```
+
+The Grafana dashboard (and API) will be accessible without requiring authentication
```

Comment:
I know it's probably because I'm too dumb, but where are the actual logs? When I open grafana on localhost:3000 I don't see any connected logs coming in?

![image](https://github.com/user-attachments/assets/35fd7442-ae0a-444a-a407-723d60306955)

This is with patch to default graph:

<details>
<summary>diff</summary>

```diff
diff --git a/resources/graphs/default.graphml b/resources/graphs/default.graphml
index 153bd52..8c276a0 100644
--- a/resources/graphs/default.graphml
+++ b/resources/graphs/default.graphml
@@ -6,12 +6,14 @@
   <key attr.name="exporter" attr.type="boolean" for="node" id="exporter"/>
   <key attr.name="collect_logs" attr.type="boolean" for="node" id="collect_logs"/>
   <key attr.name="image" attr.type="string" for="node" id="image"/>
+  <key attr.name="metrics" attr.type="string" for="node" id="metrics"/>
   <graph edgedefault="directed">
     <node id="0">
         <data key="version">27.0</data>
         <data key="bitcoin_config">-uacomment=w0</data>
         <data key="exporter">true</data>
         <data key="collect_logs">true</data>
+        <data key="metrics">blocks=getblockcount() inbounds=getnetworkinfo()["connections_in"] outbounds=getnetworkinfo()["connections_in"] mempool_size=getmempoolinfo()["size"]</data>
     </node>
     <node id="1">
         <data key="version">27.0</data>
```

</details>

Reply:
Should I see running logging containers? This is all I see:

![image](https://github.com/user-attachments/assets/225c9976-4f55-43d6-aac3-df87c0a3076d)

The script appeared to run successfully:

```log
will@ubuntu in ~/src/warnet on  rpc-gauge [$!?⇕] is 📦 v0.9.11 : 🐍 (warnet)
₿ just installlogging
resources/scripts/install_logging.sh
"grafana" already exists with the same configuration, skipping
"prometheus-community" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "grafana" chart repository
...Successfully got an update from the "prometheus-community" chart repository
Update Complete. ⎈Happy Helming!⎈
Release "loki" does not exist. Installing it now.
NAME: loki
LAST DEPLOYED: Tue Aug  6 10:11:49 2024
NAMESPACE: warnet-logging
STATUS: deployed
REVISION: 1
NOTES:
***********************************************************************
 Welcome to Grafana Loki
 Chart version: 5.47.2
 Loki version: 2.9.6
***********************************************************************

Installed components:
* gateway
* minio
* read
* write
* backend
Release "promtail" does not exist. Installing it now.
NAME: promtail
LAST DEPLOYED: Tue Aug  6 10:12:38 2024
NAMESPACE: warnet-logging
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
***********************************************************************
 Welcome to Grafana Promtail
 Chart version: 6.16.4
 Promtail version: 3.0.0
***********************************************************************

Verify the application is working by running these commands:
* kubectl --namespace warnet-logging port-forward daemonset/promtail 3101
* curl http://127.0.0.1:3101/metrics
Release "prometheus" does not exist. Installing it now.
NAME: prometheus
LAST DEPLOYED: Tue Aug  6 10:12:40 2024
NAMESPACE: warnet-logging
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
kube-prometheus-stack has been installed. Check its status by running:
  kubectl --namespace warnet-logging get pods -l "release=prometheus"

Visit https://github.com/prometheus-operator/kube-prometheus for instructions on how to create & configure Alertmanager and Prometheus instances using the Operator.
Release "loki-grafana" does not exist. Installing it now.
NAME: loki-grafana
LAST DEPLOYED: Tue Aug  6 10:12:54 2024
NAMESPACE: warnet-logging
STATUS: deployed
REVISION: 1
NOTES:
1. Get your 'admin' user password by running:

   kubectl get secret --namespace warnet-logging loki-grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo


2. The Grafana server can be accessed via port 80 on the following DNS name from within your cluster:

   loki-grafana.warnet-logging.svc.cluster.local

   Get the Grafana URL to visit by running these commands in the same shell:
     export POD_NAME=$(kubectl get pods --namespace warnet-logging -l "app.kubernetes.io/name=grafana,app.kubernetes.io/instance=loki-grafana" -o jsonpath="{.items[0].metadata.name}")
     kubectl --namespace warnet-logging port-forward $POD_NAME 3000

3. Login with the password from step 1 and the username: admin
#################################################################################
######   WARNING: Persistence is disabled!!! You will lose your data when   #####
######            the Grafana pod is terminated.                            #####
#################################################################################

will@ubuntu in ~/src/warnet on  rpc-gauge [$!?⇕] is 📦 v0.9.11 : 🐍 (warnet) took 1m8s
₿ just connectlogging
resources/scripts/connect_logging.sh
Go to http://localhost:3000
Grafana pod name: loki-grafana-6c855549d4-wsv88
Attempting to start Grafana port forwarding
Forwarding from 127.0.0.1:3000 -> 3000
Forwarding from [::1]:3000 -> 3000
Handling connection for 3000
```

Reply:
kubectl does show preometheus running:

```bash
₿ kubectl --namespace warnet-logging get pods -l "release=prometheus"
NAME                                                   READY   STATUS    RESTARTS   AGE
prometheus-kube-prometheus-operator-6c5998f7dc-hjvwx   1/1     Running   0          9m27s
prometheus-kube-state-metrics-688d66b5b8-8srsw         1/1     Running   0          9m27s
prometheus-prometheus-node-exporter-8tt7q              1/1     Running   0          9m27s
```

But I don't see any node exporters?

Reply:
server logs don't appear to contain any errors:

```log
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 0 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w0', 'tc_netem=None', 'exporter=True', 'metrics=blocks=getblockcount() inbounds=getnetworkinfo()["connections_in"] outbounds=getnetworkinfo()["connections_in"] mempool_size=getmempoolinfo()["size"]', 'collect_logs=True', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 1 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w1', 'tc_netem=None', 'exporter=True', 'metrics=None', 'collect_logs=True', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 2 with attributes: ['version=None', 'image=bitcoindevproject/bitcoin:26.0', 'bitcoin_config=-uacomment=w2 -debug=mempool', 'tc_netem=None', 'exporter=True', 'metrics=None', 'collect_logs=True', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 3 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w3', 'tc_netem=None', 'exporter=True', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 4 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w4', 'tc_netem=None', 'exporter=True', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 5 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w5', 'tc_netem=None', 'exporter=True', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 6 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w6', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 7 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w7', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 8 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w8', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 9 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w9', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 10 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w10', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | tank     | Parsed graph node: 11 with attributes: ['version=27.0', 'image=None', 'bitcoin_config=-uacomment=w11', 'tc_netem=None', 'exporter=False', 'metrics=None', 'collect_logs=False', 'build_args=', 'ln=None', 'ln_image=None', 'ln_cb_image=None', 'ln_config=None']
2024-08-06 09:11:34 | INFO    | warnet   | Imported 12 tanks from graph
warnet-rpc  | 2024-08-06 09:11:34 | INFO    | warnet   | Created Warnet using directory /root/.warnet/warnet/warnet
2024-08-06 09:11:34 | DEBUG   | k8s      | Deploying pods
warnet-rpc  | 2024-08-06 09:11:34 | DEBUG   | k8s      | Creating bitcoind container for tank 0
```

Reply:
I do see a single running exporter (I think):

```log
✗ kubectl --namespace warnet-logging get pods
NAME                                                     READY   STATUS    RESTARTS   AGE
alertmanager-prometheus-kube-prometheus-alertmanager-0   2/2     Running   0          14m
loki-backend-0                                           2/2     Running   0          16m
loki-backend-1                                           2/2     Running   0          16m
loki-backend-2                                           2/2     Running   0          16m
loki-canary-z7vgh                                        1/1     Running   0          16m
loki-gateway-6b57fdb5dd-ktspk                            1/1     Running   0          16m
loki-grafana-6c855549d4-wsv88                            1/1     Running   0          15m
loki-grafana-agent-operator-b8f4865b9-lq2fc              1/1     Running   0          16m
loki-minio-0                                             1/1     Running   0          16m
loki-read-5d8755d4cf-74zwb                               1/1     Running   0          16m
loki-read-5d8755d4cf-9wctf                               1/1     Running   0          16m
loki-read-5d8755d4cf-gdb6z                               1/1     Running   0          16m
loki-write-0                                             1/1     Running   0          16m
loki-write-1                                             1/1     Running   0          16m
loki-write-2                                             1/1     Running   0          16m
prometheus-kube-prometheus-operator-6c5998f7dc-hjvwx     1/1     Running   0          15m
prometheus-kube-state-metrics-688d66b5b8-8srsw           1/1     Running   0          15m
prometheus-prometheus-kube-prometheus-prometheus-0       2/2     Running   0          14m
prometheus-prometheus-node-exporter-8tt7q                1/1     Running   0          15m
promtail-7h8jk                                           1/1     Running   0          15m
```

Reply:
The exporters are inside the tank pods, next to the bitcoin containers.

![Screenshot 2024-08-06 at 10 27 58 AM](https://github.com/user-attachments/assets/bf4d6a7a-8118-4fc7-9a61-22dfb5e0743d)


The stuff in warnet-logging is the grafana api server and the prometheus scraper that reads from the individual tank exporters. I know there is a "node exporter" pod in warnet-logging as well, I dunno what that is actually for and on my system, it never works anyway:
![Screenshot 2024-08-06 at 10 26 52 AM](https://github.com/user-attachments/assets/7df6dcf1-b4a4-4a3a-8bd2-957a10a0e98b)

As far as seeing something in Grafana right away you're right I didn't document that, I will push another commit today that hopefully makes a default dashboard easy to load

Reply:
@willcl-ark logs are in Loki not Prometheus. There are no additional containers for Loki to get it's data as it collects it via k8s directly similar to how you can do `k logs rpd-0`.